crates/circuit/src/circuit_data.rs

```diff
@@ -160,172 +164,73 @@ impl CircuitData {
                 #[cfg(feature = "cache_pygates")]
                 py_op: RefCell::new(None),
             });
+            res.track_instruction_parameters(py, res.data.len() - 1)?;
```

Comment:
I think as long as we're accepting general `Param` instances, I think it's a mistake not to.  At the moment, the way we use it makes it very unlikely for the `Param`s to be anything other than floats, but I don't think it'd be too hard to have someone call this on a circuit they got from Python space after applying some filter / join of the iterables or something.

The `ParameterTable` isn't a public field (and imo absolutely shouldn't be), so unless the typing of this function is changed so that it only accepts `f64` as the params, I think we need to do this.

The overhead shouldn't be much, since we can determine that it's a no-op without any Python-space calls.